### PR TITLE
Plan: add typing

### PR DESCRIPTION
**Pull request by myakove**
## Summary by CodeRabbit

- **New Features**
	- Enhanced type hints for parameters in the `Plan` class constructor, improving clarity on expected input types.
	- Introduced `hooks_array` attribute to store hook specifications.
	- Added a new entry to `.gitignore` for ignoring the `.local_dev/` directory.

- **Bug Fixes**
	- Updated logic for adding hooks to virtual machines to ensure both `hooks_array` and `virtual_machines_list` are present before proceeding.

- **Changes**
	- Modified return types of the `__enter__` and `deploy` methods in the `Resource` class to `Any`, broadening expected return values.

### CODE REVIEW

**Comment by redhat-qe-bot1** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
    * You can add extra args to the Podman build command
        * Example: `/build-and-push-container --build-arg OPENSHIFT_PYTHON_WRAPPER_COMMIT=<commit_hash>`
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest tox`: Retest tox
 * `/retest python-module-install`: Retest python-module-install
 * `/retest all`: Retest all

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

File: ocp_resources/resource.py [hunk 1]
```diff
@@ -629,7 +632,7 @@ def _sigint_handler(self, signal_received: int, frame: Any) -> None:
         self.__exit__()
         sys.exit(signal_received)
 
-    def deploy(self, wait: bool = False) -> "Resource":
+    def deploy(self, wait: bool = False) -> Any:
```
[the remark this answers is not in the document]
Comment (myakove) [reply ocp_resources/resource.py:634]: The return type cannot be Resource sine each sub class has it's own name, and i's not Resource

File: ocp_resources/plan.py [hunk 2]
```diff
@@ -26,38 +26,24 @@ class Plan(NamespacedResource, MTV):
...
-        warm_migration=False,
-        pre_hook_name=None,
-        pre_hook_namespace=None,
-        after_hook_name=None,
-        after_hook_namespace=None,
-        client=None,
-        teardown=True,
-        yaml_file=None,
-        delete_timeout=TIMEOUT_4MINUTES,
+        source_provider_name: str | None = None,
```
Comment (rnetser) [inline ocp_resources/plan.py:29]: why not use the same format we use for class generator?

File: ocp_resources/plan.py [hunk 3]
```diff
@@ -139,3 +115,13 @@ def to_dict(self) -> None:
...
+
+    def _generate_hook_spec(self, hook_name: str, hook_namespace: str, hook_type: str) -> dict[str, Any]:
+        hook: dict[str, Any] = {
+            "hook": {
+                "name": hook_name,
+                "namespace": hook_namespace,
+            },
+            "step": hook_type,
+        }
+        return hook
```
Comment (rnetser) [inline ocp_resources/plan.py:127]:
```suggestion
        return {
            "hook": {
                "name": hook_name,
                "namespace": hook_namespace,
            },
            "step": hook_type,
        }
```

File: ocp_resources/plan.py (hunk 2 above)
Comment (myakove) [reply ocp_resources/plan.py:29]: what do you mean?

File: ocp_resources/plan.py (hunk 3 above)
Comment (myakove) [reply ocp_resources/plan.py]: Done

File: ocp_resources/plan.py (hunk 2 above)
Comment (rnetser) [reply ocp_resources/plan.py:29]: Using Optional and for strings would be
`source_provider_name: Optional[str] = ""`

Comment (myakove) [reply ocp_resources/plan.py:29]: `|` is the recommended way.
we should also change class-generator to use this.

Comment (myakove) [reply ocp_resources/plan.py:29]: Need to add `from __future__ import annotations`

**Review by rnetser — APPROVED** [review]

**Pull request merged**